### PR TITLE
Variant refactor

### PR DESCRIPTION
**Pull request by Vultraz**
~~Don't merge yet. Still crashes at start.~~

Main crash fixed.

### CODE REVIEW

File: src/formula/variant.cpp [hunk 1]
```diff
@@ -212,194 +188,123 @@ variant_iterator& variant_iterator::operator=(const variant_iterator& that)
...
-    : type_(TYPE_MAP)
+	: value_((std::make_shared<game_logic::variant_map>(*map)))
 {
-	assert(map);
-	map_ = new std::map<variant,variant>(*map);
+	assert(value_.get());
 }
 
 variant::variant(const variant& v)
-    : type_(v.type_)
```
Comment (CelticMinstrel) [inline src/formula/variant.cpp]: Oh huh, looks like the old copy constructor had significant memory management issues.

File: src/formula/variant.cpp [hunk 2]
```diff
@@ -212,194 +188,123 @@ variant_iterator& variant_iterator::operator=(const variant_iterator& that)
...
-	release();
 }
 
 variant& variant::operator=(const variant& v)
 {
-	if(&v != this) {
-		this->~variant();
-		new(this) variant(v);
-	}
+	value_ = v.value_;
```
Comment (CelticMinstrel) [inline src/formula/variant.cpp]: If this is all it is, you can just remove the copy constructor altogether, because this is exactly what the default compiler-generated copy constructor would do.

File: src/formula/variant.cpp [hunk 3]
```diff
@@ -212,194 +188,123 @@ variant_iterator& variant_iterator::operator=(const variant_iterator& that)
...
-		std::map<variant,variant>::const_iterator i = map_->find(v);
-		if (i == map_->end())
-		{
+	if(is_map()) {
+		auto& map = value_cast<game_logic::variant_map>()->get_container();
+
+		auto i = map.find(v);
+		if(i == map.end()) {
 			static variant null_variant;
 			return null_variant;
```
Comment (CelticMinstrel) [inline src/formula/variant.cpp]: Well, nothing wrong with this, but I wonder why it bothered. It could just `return variant()` for the same effect, and it's not like `variant` is a heavyweight class that's expensive to construct.

File: src/formula/variant.cpp [hunk 4]
```diff
@@ -410,108 +315,93 @@ variant variant::operator[](const variant& v) const
...
-	if (type_ == TYPE_LIST) {
-		assert(list_);
-		return list_->size();
-	} else if (type_ == TYPE_MAP) {
-		assert(map_);
-		return map_->size();
-	} else {
-		throw type_error(formatter() << "type error: "
-			<< " expected a list or a map but found " << type_string()
-			<< " (" << to_debug_string() << ")");
```
Comment (CelticMinstrel) [inline src/formula/variant.cpp]: So do we still get this nice error message now? I get the impression that it's gone...

File: src/formula/variant.cpp [hunk 5]
```diff
@@ -522,119 +412,99 @@ int variant::as_decimal() const
...
+		}
 
-		return variant( static_cast<int>(long_int) , variant::DECIMAL_VARIANT );
+		return variant(static_cast<int>(long_int) , variant::DECIMAL_VARIANT );
 	}
 
 	return variant(as_int() * v.as_int());
 }
 
 variant variant::operator/(const variant& v) const
```
Comment (CelticMinstrel) [inline src/formula/variant.cpp]: Seems like the multiplication and division implementations for decimals could use some work...

(Nothing to do with the variant refactor, I just happened to notice it.)

File: src/formula/variant.cpp [hunk 6]
```diff
@@ -647,121 +517,82 @@ variant variant::operator/(const variant& v) const
...
 
 variant variant::operator^(const variant& v) const
 {
-	if( type_ == TYPE_DECIMAL || v.type_ == TYPE_DECIMAL ) {
+	if(is_decimal() || v.is_decimal()) {
 
-		double res = pow( as_decimal()/1000.0 , v.as_decimal()/1000.0 );
+		double res = pow(as_decimal()/1000.0 , v.as_decimal() / 1000.0);
 
 		if(res != res) return variant();
```
Comment (CelticMinstrel) [inline src/formula/variant.cpp]: Surprised you didn't fixup the formatting here. :P

File: src/formula/variant.cpp [hunk 7]
```diff
@@ -647,121 +517,82 @@ variant variant::operator/(const variant& v) const
...
-		return *map_ == *v.map_;
-	}
-
-	case TYPE_CALLABLE: {
-		return callable_->equals(v.callable_);
-	}
-	}
-
-	return false;
+	return value_ == v.value_;
```
Comment (CelticMinstrel) [inline src/formula/variant.cpp]: Um, hold on, this is wrong. With this, two strings will compare unequal even if the coincidentally have the same value, for example.

I'm guessing what you meant is something more like `return *value_ == *v.value_`? (Did `variant_value_base` implement these operators?)

File: src/formula/variant.cpp [hunk 8]
```diff
@@ -771,57 +602,19 @@ bool variant::operator!=(const variant& v) const
 
 bool variant::operator<=(const variant& v) const
 {
-	if(type_ != v.type_) {
-		if(type_ == TYPE_DECIMAL && v.type_ == TYPE_INT) {
+	if(type() != v.type()) {
+		if(is_decimal()  && v.is_int()) {
```
Comment (CelticMinstrel) [inline src/formula/variant.cpp]: Random extra space before `&&`.

File: src/formula/variant.cpp [hunk 9]
```diff
@@ -771,57 +602,19 @@ bool variant::operator!=(const variant& v) const
...
-	}
-
-	case TYPE_CALLABLE: {
-		return !v.callable_->less(callable_);
-	}
-	}
-
-	assert(false);
-	return false;
+	return value_ <= v.value_;
```
Comment (CelticMinstrel) [inline src/formula/variant.cpp]: Again, this is pointer comparison, not value comparison. Presumably you want `*value_ <= *v.value_`.

File: src/formula/variant.cpp [hunk 10]
```diff
@@ -841,97 +634,49 @@ bool variant::operator>(const variant& v) const
 
 variant variant::list_elements_add(const variant& v) const
 {
-	must_be(TYPE_LIST);
-	v.must_be(TYPE_LIST);
-
-	if( num_elements() != v.num_elements() )
-		throw type_error("Operator '.+' requires two lists of the same length");
```
Comment (CelticMinstrel) [inline src/formula/variant.cpp]: So I guess we've lost this nice error message? At least, the mention of the specific operator in it.

File: src/formula/variant.cpp [hunk 11]
```diff
@@ -841,97 +634,49 @@ bool variant::operator>(const variant& v) const
...
-	res.reserve(num_elements());
-
-	for(size_t i = 0; i < num_elements(); ++i) {
-		res.push_back( (*this)[i] / v[i] );
-	}
-
-	return variant( &res );
+	must_both_be(VARIANT_TYPE::TYPE_LIST, v);
+	return value_cast<game_logic::variant_list>()->list_op(v.value_, [this](variant& v1, variant& v2) { return v1 / v2; });
 }
```
Comment (CelticMinstrel) [inline src/formula/variant.cpp]: Hmm, now I'm wondering why there's no `.^` operator – though I guess the use for that would be rather low – or `.%` operator.

File: src/formula/variant.cpp [hunk 12]
```diff
@@ -941,141 +686,54 @@ variant variant::concatenate(const variant& v) const
...
-		assert(false);
+	if(type() != t || second.type() != t) {
+		throw type_error(formatter() << "type error: expected "
+			<< variant_type_to_string(t) << " but found "
+			<<        type_string() << " (" <<        to_debug_string() << ")" << " and "
+			<< second.type_string() << " (" << second.to_debug_string() << ")");
 	}
 }
 
+std::string variant::serialize_to_string() const
```
Comment (CelticMinstrel) [inline src/formula/variant.cpp]: Again... why did you change this API to return the string instead of taking it as a reference?

File: src/formula/callable.hpp [hunk 13]
```diff
@@ -117,8 +119,6 @@ class formula_callable {
 	}
 
 protected:
-	virtual ~formula_callable() {}
```
Comment (Vultraz) [inline src/formula/callable.hpp]: It was just moved up.

File: src/formula/callable_fwd.hpp [hunk 14]
```diff
@@ -15,13 +15,17 @@
...
+namespace game_logic
+{
 class formula_callable;
 class formula_debugger;
-typedef std::shared_ptr<formula_callable> formula_callable_ptr;
-typedef std::shared_ptr<const formula_callable> const_formula_callable_ptr;
 
+using formula_callable_ptr       = std::shared_ptr<formula_callable>;
+using const_formula_callable_ptr = std::shared_ptr<const formula_callable>;
+using const_formula_callable_vec = std::vector<const formula_callable*>; // TODO: use shared_ptr
```
Comment (Vultraz) [inline src/formula/callable_fwd.hpp]: `to_debug_string` impl.

File: src/formula/function.cpp [hunk 15]
```diff
@@ -1479,9 +1479,9 @@ void key_value_pair::get_inputs(formula_input_vector* inputs) const {
 
 void key_value_pair::serialize_to_string(std::string& str) const {
 	str += "pair(";
-	key_.serialize_to_string(str);
+	str += key_.serialize_to_string();
```
Comment (Vultraz) [inline src/formula/function.cpp]: It was previously taking the string by reference and using `+=` widely.

File: src/formula/variant.hpp [hunk 16]
```diff
@@ -15,41 +15,40 @@
...
-	~call_stack_manager() {
+	~call_stack_manager()
+	{
 		pop_call_stack();
 	}
 };
 
 class variant_iterator;
 
-struct type_error : public game::error {
```
Comment (Vultraz) [inline src/formula/variant.hpp]: I've moved it.

File: src/formula/variant_private.cpp [hunk 17]
```diff
@@ -0,0 +1,282 @@
...
+}
+
+variant_callable::variant_callable(const formula_callable* callable)
+	: callable_(callable)
+	, mutable_callable_(nullptr) // FIXME
+{}
+
+std::string variant_callable::get_serialized_string() const
+{
+	// TODO: make serialize return a string.
```
Comment (Vultraz) [inline src/formula/variant_private.cpp]: I might as well change all serialize functions to return strings not just one.

File: src/formula/callable.hpp (hunk 13 above)
Comment (CelticMinstrel) [inline src/formula/callable.hpp]: Ohhhhh, sorry, I guess I missed it. >_>

File: src/formula/callable_fwd.hpp (hunk 14 above)
Comment (CelticMinstrel) [inline src/formula/callable_fwd.hpp]: Oh, for "seen"? Does it really need to be public, then? It could be defined in `variant_callable` for example.

File: src/formula/variant.hpp (hunk 16 above)
Comment (CelticMinstrel) [inline src/formula/variant.hpp]: Yeah, I noticed that later, sorry.

File: src/formula/variant.cpp (hunk 7 above)
Comment (Vultraz) [inline src/formula/variant.cpp]: variant_value_base implements == and <= virtually, yes.

File: src/formula/variant_private.cpp (hunk 17 above)
Comment (CelticMinstrel) [inline src/formula/variant_private.cpp]: I think it would be better to make formula_callable::serialize take the stringstream as a parameter (specifically, probably as an `std::ostream`).

File: src/formula/variant_private.cpp [hunk 18]
```diff
@@ -0,0 +1,282 @@
...
+		if(!first_time) {
+			ss << ",";
+		}
+
+		first_time = false;
+
+		ss << to_string_detail(member, mod_func);
+	}
+
+	// TODO: evaluate if this really needs to be separately conditional.
```
Comment (CelticMinstrel) [inline src/formula/variant_private.cpp]: And it didn't appear in all places the [] appeared?

File: src/formula/variant_private.cpp [hunk 19]
```diff
@@ -0,0 +1,282 @@
...
+
+	return true;
+}
+
+bool variant_list::operator<=(value_base_ptr other) const
+{
+	const auto& other_container = value_cast<variant_list>(other)->v_vector_;
+
+	for(size_t n = 0; n != num_elements() && n != other->num_elements(); ++n) {
+		// TODO: shouldn't this just be <=?
```
Comment (Vultraz) [inline src/formula/variant_private.cpp]: No, I meant `<=` for each element. The code checks for `<` on the current two elements. then `>` on the current two elements. It seems weird, since it looks like `==` is never considered.

File: src/formula/variant_private.hpp [hunk 20]
```diff
@@ -0,0 +1,457 @@
...
+ *
+ * Do note this class should implement *no* data members.
+ */
+class variant_value_base
+{
+public:
+	/** Returns the number of elements in a type. Not relevant for every derivative. */
+	virtual size_t num_elements() const
+	{
+		return 0;
```
Comment (Vultraz) [inline src/formula/variant_private.hpp]: Only for callable. It was an error otherwise.

File: src/formula/variant_private.hpp [hunk 21]
```diff
@@ -0,0 +1,457 @@
...
+	}
+
+	virtual std::string get_serialized_string() const
+	{
+		return "null()";
+	}
+
+	virtual std::string get_debug_string(const_formula_callable_vec* /*seen*/, bool /*verbose*/) const
+	{
+		return get_serialized_string();
```
Comment (Vultraz) [inline src/formula/variant_private.hpp]: Ah, right, it was.

File: src/formula/variant_private.hpp [hunk 22]
```diff
@@ -0,0 +1,457 @@
...
+	virtual std::string string_cast() const override
+	{
+		return string_;
+	}
+
+	virtual std::string get_serialized_string() const override;
+
+	virtual std::string get_debug_string(const_formula_callable_vec* /*seen*/, bool /*verbose*/) const override
+	{
+		return string_;
```
Comment (Vultraz) [inline src/formula/variant_private.hpp]: Ah, indeed, it was wrapped in `''` before.

File: src/formula/variant_private.cpp (hunk 18 above)
Comment (Vultraz) [inline src/formula/variant_private.cpp]: Only the `serialize_to_string` impl.

File: src/formula/variant.cpp [hunk 23]
```diff
@@ -12,40 +12,23 @@
...
-	case variant::TYPE_STRING:
-		return "string";
-	case variant::TYPE_MAP:
-		return "map";
-	default:
-		assert(false);
-		return "invalid";
-	}
+
+std::string variant_type_to_string(VARIANT_TYPE type)
```
Comment (Vultraz) [inline src/formula/variant.cpp]: Yes.

File: src/formula/variant_private.cpp (hunk 19 above)
Comment (CelticMinstrel) [inline src/formula/variant_private.cpp]: `==` is considered implicitly. If two elements are equal, it needs to check the next pair of elements to determine the result. Thus, the proper behaviour when they're equal is to continue to the next iteration of the loop.

File: src/formula/variant.cpp (hunk 7 above)
Comment (CelticMinstrel) [inline src/formula/variant.cpp]: (Kinda wonder if it'd be better to name them instead of using operators, though.)

File: src/formula/variant_private.hpp [hunk 24]
```diff
@@ -0,0 +1,466 @@
...
+};
+
+
+template<typename T>
+class variant_container : public virtual variant_value_base
+{
+public:
+	explicit variant_container(const T& container)
+		: container_(container)
+		, container_iter_(container_.begin())
```
Comment (CelticMinstrel) [inline src/formula/variant_private.hpp]: Pretty sure there's no guarantee that `begin()` will be valid after inserting some elements. (Possibly, even after removing some.) So uh... it's probably a bad idea to be actually storing it like this.

File: src/formula/variant.hpp [hunk 25]
```diff
@@ -128,8 +136,8 @@ class variant {
 
 	bool operator==(const variant&) const;
 	bool operator!=(const variant&) const;
-	bool operator<(const variant&) const;
-	bool operator>(const variant&) const;
+	bool operator<(const variant&)  const;
```
Comment (Vultraz) [inline src/formula/variant.hpp]: I wanted the `const`s to line up.

File: src/formula/variant.cpp (hunk 4 above)
Comment (Vultraz) [inline src/formula/variant.cpp]: It's gone.

File: src/formula/variant.cpp (hunk 10 above)
Comment (Vultraz) [inline src/formula/variant.cpp]: Specific op is gone, message is still around.

File: src/formula/variant.cpp [hunk 26]
```diff
@@ -522,119 +411,99 @@ int variant::as_decimal() const
...
 {
-	if(type_ == TYPE_LIST) {
-		if(v.type_ == TYPE_LIST) {
-			std::vector<variant> res;
-			res.reserve(list_->size() + v.list_->size());
-			for(size_t i = 0; i<list_->size(); ++i) {
-				const variant& var = (*list_)[i];
-				res.push_back(var);
-			}
+	if(is_list() && v.is_list()) {
```
Comment (Vultraz) [inline src/formula/variant.cpp]: What?

File: src/formula/variant_private.hpp [hunk 27]
```diff
@@ -0,0 +1,466 @@
...
+	}
+
+	virtual std::string get_serialized_string() const
+	{
+		return "null()";
+	}
+
+	virtual std::string get_debug_string(const_formula_callable_vec& /*seen*/, bool /*verbose*/) const
+	{
+		return get_serialized_string();
```
Comment (Vultraz) [inline src/formula/variant_private.hpp]: Uh... it doesn't matter to me either way. Which would you prefer?

File: src/formula/variant.hpp (hunk 25 above)
Comment (CelticMinstrel) [inline src/formula/variant.hpp]: ...there should be `operator>`, `operator<=` and `operator>=` here...

File: src/formula/variant.cpp (hunk 26 above)
Comment (CelticMinstrel) [inline src/formula/variant.cpp]: Concatenation of lists with + rather than ..

File: src/formula/variant_private.hpp [hunk 28]
```diff
@@ -54,6 +54,9 @@ class variant_value_base;
 
 using value_base_ptr = std::shared_ptr<variant_value_base>;
 
+/**
+ * Helper functions to cast a @ref variant_value_base to a new derived type.
```
Comment (CelticMinstrel) [inline src/formula/variant_private.hpp]: You can't use one doc comment for two functions like this. Make "function" singular and duplicate the comment for each function.

If the duplication _really_ bugs you though, there _is_ a [`@copydoc` command](https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdcopydoc), but that might be a little difficult to use with a template function.

File: src/formula/variant.hpp (hunk 25 above)
Comment (Vultraz) [inline src/formula/variant.hpp]: There is?

**Comment by CelticMinstrel** [comment]
Alright, now that the null pointers are gone, I can approve this. However, there are still a number of my comments that you didn't address (not all of equal priority, mind you); I'll summarize here for you. The ones containing bold text, I consider important. Most are low priority.

* You haven't explained yet why `const_formula_callable_vec` is public.
* The `.get()`s strewn around in several places should be removed at some point (most likely by propagating the `shared_ptr`s into everywhere), though this is low priority, so feel free to ignore it for now.
* I'm still not happy with the filanames _variant_private.?pp_, for two reasons. First, _variant.hpp_ includes _variant_private.hpp_, which should not happen. Second, rather than "variant" they're all about "variant values", so I think maybe _variant_value.?pp_ or _variant_data.?pp_ might fit better.
* There's a code comment on line 159 or so of _variant.hpp_ containing `decltype` which should probably be removed.
* The doc comment for `variant::value_` is inaccurate. The constructors don't do any casting.
* While not high priority, it might be nice to make the serialization stream simply pass around an `ostream`.
* In `variant_iterator::operator==` I think it would look nicer not to use the ternary operator - I mentioned a different expression earlier which should give the same results.
* You no longer need an explicit copy constructor in `variant` - you can remove it and let the compiler generate a default one.
* That one error message about expecting a list or map should probably be restored somehow.
* I think the error message in `must_both_be` could be improved slightly.
* You shouldn't store a container iterator in `variant_container`. This one is rather important, actually.

**Comment by CelticMinstrel** [comment]
I think the only things you haven't addressed now (either in a commit or as a response) are the `decltype` comment (though I might've missed it) and the unnecessary `variant` copy constructor.
